8354920: SA core file support on Linux only prints error messages when debug logging is enabled

djelinski opened the pull request:
Currently if loading a core file fails, the diagnostic information provided on different systems is different; on MacOS we produce an error message, while on Linux we only print information if debug logging is enabled.

This PR adds some new messages on Linux to match MacOSX, and changes some of the diagnostic output to error level instead of debug. Additionally, if opening the core or the exe file fails, the system error message (strerror) is printed.

Tier1-3 testing clean.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8354920](https://bugs.openjdk.org/browse/JDK-8354920): SA core file support on Linux only prints error messages when debug logging is enabled (**Bug** - P4)


### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24722/head:pull/24722` \
`$ git checkout pull/24722`

Update a local copy of the PR: \
`$ git checkout pull/24722` \
`$ git pull https://git.openjdk.org/jdk.git pull/24722/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24722`

View PR using the GUI difftool: \
`$ git pr show -t 24722`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24722.diff">https://git.openjdk.org/jdk/pull/24722.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24722#issuecomment-2812733154)
</details>

plummercj reviewed (changes requested):
This is a good start for getting a reasonable error message without having to first enable LIBSAPROC_DEBUG. A couple of suggestions for additional improvements below:

You've covered the top level error message in Pgrab_core(), but there are many other print_debug messages that are for errors that are closer to the root cause of the error. For example, look in core_handle_note() and read_lib_info(). Also, Pgrab() needs updating, although it only has one debug message, but it also is missing detecting when add_thread_info() or read_lib_info() fails.

add_map_info() and core_handle_prstatus() are lacking a print_error message when they fail.

djelinski commented:
Thanks @plummercj for the review!

I added a few more print_error messages, converted some printf and print_debug to print_error, and added a new print_debug log to add_class_share_map_info, because the failure of that method is not fatal. Let me know if that looks better.

plummercj reviewed (approved):
Overall this looks great. Something I've wanted to do for a long time now but have never gotten around to. There are some places where I think you end up doubling up on error messages. For example, add_map_info() now prints an error message and so do its callers, but that's probably ok.

Please run all the SA tests in test/hotspot/jtreg/serviceability/sa and test/jdk/sun/tools/jhsdb on all our supported platforms.

Also test by hand attaching to both a core file and a process just to make sure you haven't turned any expected debug messages into error messages.

Thanks!

kevinjwalls commented on src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c:
```diff
@@ -314,8 +318,10 @@ static bool read_core_segments(struct ps_prochandle* ph, ELF_EHDR* core_ehdr) {
    ELF_PHDR* core_php = NULL;
 
-   if ((phbuf =  read_program_header_table(ph->core->core_fd, core_ehdr)) == NULL)
+   if ((phbuf =  read_program_header_table(ph->core->core_fd, core_ehdr)) == NULL) {
```
odd double space

kevinjwalls commented on src/jdk.hotspot.agent/linux/native/libsaproc/ps_proc.c:
```diff
@@ -447,29 +447,38 @@ Pgrab(pid_t pid, char* err_buf, size_t err_buf_len) {
   if ( (ph = (struct ps_prochandle*) calloc(1, sizeof(struct ps_prochandle))) == NULL) {
     snprintf(err_buf, err_buf_len, "can't allocate memory for ps_prochandle");
-    print_debug("%s\n", err_buf);
+    print_error("%s\n", err_buf);
```
This might be one that would print it twice - the caller in LinuxDebuggerLocal.cpp will print what's in err_buf.

kevinjwalls replied:
Looks like here in Pgrab the intent is to snprintf an error message in err_buf, and the caller prints it.  So Pgrab should not need the print_error calls.

Some functions which are called, like read_lib_info, may print their own specific error message, then Pgrab stores an error that "failed to read lib info".. These are reporting the same thing but I don't see that as a problem or duplication, it could be really useful to track where things fail.

Yes it's a little annoying that Pgrab_core is different, it doesn't take an error_buf so needs to print_error about any problems.

djelinski replied:
ah, interesting. The native caller doesn't print the error message, it throws a Java exception with the message in the exception string, and the Java code then deals with the exception.

This is different on MacOS, where the error message is printed, and then an exception with a generic message is thrown. I haven't checked what the Windows implementation does here.

The Linux implementation with dynamic exception messages was introduced in 57d8a71115b8fc9a2eb2be876a396c474c207cf3.

I'll modify the Linux Pgrab to report all errors through the exception message, and remove print_error from Pgrab.

Should I additionally modify the MacOS's Java_sun_jvm_hotspot_debugger_bsd_BsdDebuggerLocal_attach0__I to report all errors through the exception message and remove print_error, or is it OK if the implementations behave differently?

kevinjwalls replied:
> I'll modify the Linux Pgrab to report all errors through the exception message, and remove print_error from Pgrab.
That would be great.  There is some existing fragmentation/platform differences here, so I don't think you need to make all the implementations the same for this change.

djelinski replied:
I ended up not removing print_error; I found that ptrace_attach was reporting the failures using both print_error and the exception message, so it looked reasonable to use both mechanisms in other places too.

kevinjwalls replied:
ok sounds fair!

djelinski commented:
> Please run all the SA tests in test/hotspot/jtreg/serviceability/sa and test/jdk/sun/tools/jhsdb on all our supported platforms.

Local testing on Linux completed successfully; had to reconfigure ptrace, otherwise many tests were skipped. Mach5 still in progress.

> Also test by hand attaching to both a core file and a process just to make sure you haven't turned any expected debug messages into error messages.

Done (Linux only, I don't have access to a Mac). I didn't observe any unintended error messages. Also, I only added print_error in places where it guaranteed a failure of Pgrab or Pgrab_core.

plummercj reviewed (approved):
Looks good. Thanks for all the added cleanup.

djelinski commented:
Thanks for the reviews! Mach5 testing came back clean.

/integrate
